Project/Source/Components/ComponentParticleSystem.cpp

```diff
@@ -0,0 +1,305 @@
+		float z = (transform->GetGlobalPosition().z) + (float(rand()) / float((RAND_MAX)) * 0.2) - 0.2;
+		float y = (transform->GetGlobalPosition().y);
+		return float3(x, y, z);
+	}
+	if (emitterType == EmitterType::SPHERE) {
+		ComponentTransform* transform = GetOwner().GetComponent<ComponentTransform>();
+		float x = (transform->GetGlobalPosition().x) + (float(rand()) / float((RAND_MAX)) * 0.5) - 0.5;
+		float z = (transform->GetGlobalPosition().z) + (float(rand()) / float((RAND_MAX)) * 0.5) - 0.5;
+		float y = (transform->GetGlobalPosition().y) + (float(rand()) / float((RAND_MAX)) * 0.5) - 0.5;
+		return float3(transform->GetGlobalPosition());
+	}
+}
+
+void ComponentParticleSystem::CreateParticles(float nParticles, float vel) {
+	particles.Allocate(maxParticles);
+	for (Particle& currentParticle : particles) {
+		currentParticle.scale = float3(0.1f, 0.1f, 0.1f) * scale;
+		currentParticle.initialPosition = CreatePosition();
+		currentParticle.position = currentParticle.initialPosition;
+		currentParticle.direction = CreateVelocity();
+		currentParticle.velocity = vel;
+	}
+}
+
+void ComponentParticleSystem::Load(JsonValue jComponent) {
+	shaderID = jComponent[JSON_TAG_TEXTURE_SHADERID];
+
+	if (shaderID != 0) {
+		App->resources->IncreaseReferenceCount(shaderID);
+	}
+
+	textureID = jComponent[JSON_TAG_TEXTURE_TEXTUREID];
+
+	if (textureID != 0) {
+		App->resources->IncreaseReferenceCount(textureID);
+	}
+
+	CreateParticles(maxParticles, velocity);
+}
+
+void ComponentParticleSystem::Save(JsonValue jComponent) const {
+	jComponent[JSON_TAG_TEXTURE_SHADERID] = shaderID;
+	jComponent[JSON_TAG_TEXTURE_TEXTUREID] = textureID;
+}
+
+void ComponentParticleSystem::Update() {
+	deadParticles.clear();
+	for (Particle& currentParticle : particles) {
+		currentParticle.life -= 0.01;
+		currentParticle.position += currentParticle.direction * velocity;
+		currentParticle.model = float4x4::FromTRS(currentParticle.position, currentParticle.rotation, currentParticle.scale);
+		if (currentParticle.life < 0) {
+			deadParticles.push_back(&currentParticle);
+		}
+	}
+	for (Particle* currentParticle : deadParticles) {
+		particles.Release(currentParticle);
+	}
+	Draw();
+}
+
+void ComponentParticleSystem::Init() {
+	CreateParticles(maxParticles, velocity);
+}
+
+void ComponentParticleSystem::SpawnParticle() {
+	Particle* currentParticle = particles.Obtain();
+	if (currentParticle) {
+		currentParticle->position = currentParticle->initialPosition;
+		currentParticle->life = particleLife;
+		if (isRandomFrame) {
+			currentParticle->currentFrame = rand() % ((Xtiles * Ytiles) + 1);
+		} else {
+			currentParticle->currentFrame = 0;
+		}
+		currentParticle->colorFrame = 0;
+		currentParticle->direction = CreateVelocity();
+		currentParticle->initialPosition = CreatePosition();
+		currentParticle->position = currentParticle->initialPosition;
+		currentParticle->scale = float3(0.1f, 0.1f, 0.1f) * scale;
```

Comment:
Hardcoded values should be maked with a TODO so that we know what to change later.